cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -0,0 +1,381 @@
+    }
+  }
+
+  /**
+   * Returns the transaction for the Spark program execution.
+   */
+  public Transaction getTransaction() {
+    return transaction;
+  }
+
+  /**
+   * Returns the {@link ApacheSparkContext}.
+   *
+   * @throws IllegalStateException if the {@link ApacheSparkContext} is not yet set.
+   */
+  private synchronized ApacheSparkContext getApacheSparkContext() {
+    Preconditions.checkState(apacheSparkContext != null,
+                             "Spark Context not available in the current execution context");
+    return apacheSparkContext;
+  }
+
+  /**
+   * Calls {@link TransactionAware#commitTx()} on the given dataset followed by a {@link Closeable#close()} call.
+   */
+  private void commitAndClose(Dataset dataset) {
+    try {
+      if (dataset instanceof TransactionAware) {
+        try {
+          ((TransactionAware) dataset).commitTx();
+        } catch (Exception e) {
+          // Nothing much we can do besides propagating it.
+          // Since the propagation will end up in the Spark Framework, we do the logging here.
+          LOG.error("Failed to commit dataset changes", e);
+          throw Throwables.propagate(e);
+        }
+      }
+    } finally {
+      Closeables.closeQuietly(dataset);
+    }
+  }
+
+  /**
+   * Setups the configuration for reading from stream through {@link StreamInputFormat}.
+   */
+  private Configuration configureStreamInput(Configuration configuration,
+                                             StreamBatchReadable stream, Class<?> vClass) throws IOException {
+
+    Id.Stream streamId = Id.Stream.from(getProgramId().getNamespace(), stream.getStreamName());
+
+    StreamConfig streamConfig = streamAdmin.getConfig(streamId);
+    Location streamPath = StreamUtils.createGenerationLocation(streamConfig.getLocation(),
+                                                               StreamUtils.getGeneration(streamConfig));
+    StreamInputFormat.setTTL(configuration, streamConfig.getTTL());
+    StreamInputFormat.setStreamPath(configuration, streamPath.toURI());
+    StreamInputFormat.setTimeRange(configuration, stream.getStartTime(), stream.getEndTime());
+
+    String decoderType = stream.getDecoderType();
+    if (decoderType == null) {
+      // If the user don't specify the decoder, detect the type
+      StreamInputFormat.inferDecoderClass(configuration, vClass);
+    } else {
+      StreamInputFormat.setDecoderClassName(configuration, decoderType);
+    }
+    return configuration;
+  }
+
+  /**
+   * Creates a new instance of dataset.
+   */
+  private <T extends Dataset> T createDataset(String datasetName, Map<String, String> arguments) {
+    T dataset = datasetInstantiator.getDataset(datasetName, arguments);
+
+    // Provide the current long running transaction to the given dataset.
+    // Remove it from the transaction aware list from the dataset instantiator, since we won't use it in this context
+    // Need to remove it to avoid memory leak.
```

Comment:
"This context" in the comment I refer to the context when this class (`ExecutionSparkContext`) is use.